File: cmd/neuratrade-cli/local_config.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

type localConfig struct {
	TelegramTestChatID string `json:"telegram_test_chat_id"`
	Server struct {
		Host string `json:"host"`
		Port int    `json:"port"`
	} `json:"server"`
	Database struct {
		SQLitePath string `json:"sqlite_path"`
	} `json:"database"`
	CCXT struct {
		AdminAPIKey string `json:"admin_api_key"`
		ServiceURL  string `json:"service_url"`
		GrpcAddress string `json:"grpc_address"`
	} `json:"ccxt"`
	Telegram struct {
		BotToken   string `json:"bot_token"`
		ApiBaseURL string `json:"api_base_url"`
		ServiceURL string `json:"service_url"`
		ChatID     string `json:"chat_id"`
	} `json:"telegram"`
	Services struct {
		Telegram struct {
			ChatID string `json:"chat_id"`
		} `json:"telegram"`
	} `json:"services"`
	Security struct {
		AdminAPIKey string `json:"admin_api_key"`
	} `json:"security"`
	AI struct {
		APIKey   string `json:"api_key"`
		BaseURL  string `json:"base_url"`
		Provider string `json:"provider"`
		Model    string `json:"model"`
	} `json:"ai"`
}

func defaultNeuraTradeHome() string {
	if v := os.Getenv("NEURATRADE_HOME"); v != "" {
		return v
	}
	return os.ExpandEnv("${HOME}/.neuratrade")
}

func loadLocalConfig(home string) (*localConfig, error) {
	configPath := filepath.Join(home, "config.json")
	content, err := os.ReadFile(configPath)
	if err != nil {
		return nil, fmt.Errorf("read config: %w", err)
	}

	var cfg localConfig
	if err := json.Unmarshal(content, &cfg); err != nil {
		return nil, fmt.Errorf("parse config: %w", err)
	}

	return &cfg, nil
}

func getConfigValue(home string) *localConfig {
	cfg, err := loadLocalConfig(home)
	if err != nil {
		return nil
	}
	return cfg
}

func configAdminAPIKey(cfg *localConfig) string {
	if cfg == nil {
		return ""
	}
	if cfg.Security.AdminAPIKey != "" {
		return cfg.Security.AdminAPIKey
	}
	return cfg.CCXT.AdminAPIKey
}

func configChatID(cfg *localConfig) string {
	if cfg == nil {
		return ""
	}
	if cfg.Telegram.ChatID != "" {
		return cfg.Telegram.ChatID
	}
	if cfg.Services.Telegram.ChatID != "" {
		return cfg.Services.Telegram.ChatID
	}
	return cfg.TelegramTestChatID
}

File: services/neuratrade-cli-ts/src/schemas/local-config.ts
/**
 * Schema definitions for localConfig (config.json).
 *
 * Mirrors the Go struct in cmd/neuratrade-cli/local_config.go.
 * All string fields are optional — the Go loader treats missing/empty as ""
 * and resolves values via fallback chains at runtime.
 */
import * as S from "effect/Schema";

// -- Nested sub-schemas --

const AuthConfig = S.Struct({
  jwt_secret: S.optional(S.String),
});

const ServerConfig = S.Struct({
  host: S.optional(S.String),
  port: S.optional(S.Number),
});

const DatabaseConfig = S.Struct({
  driver: S.optional(S.String),
  sqlite_path: S.optional(S.String),
});

const CCXTConfig = S.Struct({
  admin_api_key: S.optional(S.String),
  service_url: S.optional(S.String),
  grpc_address: S.optional(S.String),
});

const TelegramConfig = S.Struct({
  bot_token: S.optional(S.String),
  api_base_url: S.optional(S.String),
  service_url: S.optional(S.String),
  grpc_address: S.optional(S.String),
  chat_id: S.optional(S.String),
});

const ServicesTelegramConfig = S.Struct({
  chat_id: S.optional(S.String),
});

const ServicesConfig = S.Struct({
  telegram: S.optional(ServicesTelegramConfig),
});

const SecurityConfig = S.Struct({
  admin_api_key: S.optional(S.String),
  jwt_secret: S.optional(S.String),
});

const AIConfig = S.Struct({
  api_key: S.optional(S.String),
  base_url: S.optional(S.String),
  provider: S.optional(S.String),
  model: S.optional(S.String),
});

// -- Main localConfig schema --

export const LocalConfigSchema = S.Struct({
  admin_api_key: S.optional(S.String),
  telegram_test_chat_id: S.optional(S.String),
  auth: S.optional(AuthConfig),
  server: S.optional(ServerConfig),
  database: S.optional(DatabaseConfig),
  ccxt: S.optional(CCXTConfig),
  telegram: S.optional(TelegramConfig),
  services: S.optional(ServicesConfig),
  security: S.optional(SecurityConfig),
  ai: S.optional(AIConfig),
});

export type LocalConfig = typeof LocalConfigSchema.Type;

/** Decode an unknown JSON value into LocalConfig (returns Effect). */
export const decodeLocalConfig = S.decodeUnknown(LocalConfigSchema);

/** Decode an unknown JSON value, returning Either. */
export const decodeLocalConfigEither = S.decodeUnknownEither(LocalConfigSchema);
